feat(VoteProgram): support VoteInstruction::Authorize

mooori opened the pull request:
#### Problem

Solana-web3.js does not support [`VoteInstruction::Authorize`](https://docs.rs/solana-vote-program/1.9.5/solana_vote_program/vote_instruction/enum.VoteInstruction.html#variant.Authorize).

#### Summary of Changes
Support it by adding a constructor and decoder for `VoteInstruction::Authorize`.

Fixes feature request in [#22932 (comment)](https://github.com/solana-labs/solana/pull/22932#pullrequestreview-873547038).

CriesofCarrots commented on web3.js/test/vote-program.test.ts:
```diff
@@ -154,13 +173,57 @@ describe('VoteProgram', () => {
+        preflightCommitment: 'confirmed',
+      });
+
+      // Test old authorized cannot witdraw anymore.
```
```suggestion
      // Test old authorized cannot withdraw anymore.
```

CriesofCarrots commented on web3.js/test/vote-program.test.ts:
```diff
@@ -154,13 +173,57 @@ describe('VoteProgram', () => {
+        sendAndConfirmTransaction(connection, withdraw, [authorized], {
+          preflightCommitment: 'confirmed',
+        }),
+      ).to.be.rejected;
```
Do you mind adding one more transaction to demonstrate a successful withdraw with the new authority?

CriesofCarrots commented on web3.js/test/vote-program.test.ts:
```diff
@@ -154,13 +173,57 @@ describe('VoteProgram', () => {
         LAMPORTS_PER_SOL,
       );
+
+      const newAuthorized = Keypair.generate();
```
Can you also add a case to test that the authorized withdrawer can sign to authorize a new Voter? More of a demo to the user than testing anything in your implementation.
ref: https://github.com/solana-labs/solana/pull/22522
Might need to set up separate newVoter and newWithdrawer Keypairs here.

CriesofCarrots reviewed (commented):
Impl looks great! One typo nit, plus a couple cases for the integration test that I think would be helpful, if you're willing?

mooori commented:
I've extended the integration test and fixed the typo.

When trying to authorize a new Voter more than once per epoch, the transaction fails with [`VoteError::TooSoonToReauthorize`](https://docs.rs/solana-vote-program/1.9.5/solana_vote_program/vote_instruction/enum.VoteError.html#variant.TooSoonToReauthorize).

To let Witdrawer authorize a new Voter, I removed the test of Voter authorizing a new Voter (as it's quite straighforward). Alternatively, to keep that test, a new Vote account could be created.

mooori commented:
I think the CI failure (except web3-commit-lint) is not related to this PR and locally `test:live-with-test-validator` passes.

@CriesofCarrots can you see if there's something wrong on my side or if it's an issue with CI?

CriesofCarrots commented:
> When trying to authorize a new Voter more than once per epoch, the transaction fails with [`VoteError::TooSoonToReauthorize`](https://docs.rs/solana-vote-program/1.9.5/solana_vote_program/vote_instruction/enum.VoteError.html#variant.TooSoonToReauthorize).

Ah yeah, true. Probably not worth mucking with the epoch or adding a second test for that. What you've done seems fine to me, thanks!

> can you see if there's something wrong on my side or if it's an issue with CI?

At a glance, it doesn't look like anything to do with your changes. I'll take a closer look and get back to you. May ask you to rebase, but otherwise this seems gtg.

CriesofCarrots commented:
@mooori , should be fixed now. Can you rebase and push plz?

mooori commented:
> @mooori , should be fixed now. Can you rebase and push plz?

Done and now CI passed :)

Thank you!

CriesofCarrots reviewed (approved):
Thank you!
